drivers/serial/uart_nrf_sw_lpuart.c

```diff
@@ -567,6 +608,206 @@ static int api_rx_disable(struct device *dev)
 	return uart_rx_disable(data->uart);
 }
 
+static uint32_t int_driven_rd_available(struct lpuart_data *data)
+{
+	return data->int_driven->rxlen - data->int_driven->rxrd;
+}
+
+static void int_driven_rx_feed(struct device *dev, struct lpuart_data *data)
+{
+	int err;
+
+	data->int_driven->rxlen = 0;
+	data->int_driven->rxrd = 0;
+	err = api_rx_buf_rsp(dev, data->int_driven->rxbuf,
+			     sizeof(data->int_driven->rxbuf));
+	__ASSERT_NO_MSG(err >= 0);
+}
+
+static void int_driven_evt_handler(struct device *uart,
+				   struct uart_event *evt,
+				   void *user_data)
+{
+	struct device *dev = user_data;
+	struct lpuart_data *data = get_dev_data(dev);
+	bool call_handler = false;
+
+	switch (evt->type) {
+	case UART_TX_DONE:
+		data->int_driven->txlen = 0;
+		call_handler = true;
+		break;
+	case UART_RX_RDY:
+		__ASSERT_NO_MSG(data->int_driven->rxlen == 0);
+		data->int_driven->rxlen = evt->data.rx.len;
+		call_handler = data->int_driven->rx_enabled;
+		break;
+	case UART_RX_BUF_REQUEST:
+		if (int_driven_rd_available(data) == 0) {
+			int_driven_rx_feed(dev, data);
+		}
+		break;
+	case UART_RX_STOPPED:
+		call_handler = data->int_driven->err_enabled;
+		break;
+	case UART_RX_DISABLED:
+	{
+		int err;
+
+		data->int_driven->rxlen = 0;
+		data->int_driven->rxrd = 0;
+		err = api_rx_enable(dev, data->int_driven->rxbuf,
+					sizeof(data->int_driven->rxbuf), 1);
+		__ASSERT_NO_MSG(err >= 0);
+		break;
+	}
+	default:
+		break;
+	}
+
+	if (call_handler) {
+		data->int_driven->callback(dev, data->int_driven->user_data);
+	}
+}
+
+static int api_fifo_read(struct device *dev, uint8_t *rx_data, const int size)
+{
+	struct lpuart_data *data = get_dev_data(dev);
+	uint32_t available = int_driven_rd_available(data);
+	uint32_t cpylen = 0;
+
+	if (available) {
+		cpylen = MIN(available, size);
+		memcpy(rx_data,
+		       &data->int_driven->rxbuf[data->int_driven->rxrd],
+		       cpylen);
+		data->int_driven->rxrd += cpylen;
+		if ((data->rx_state == RX_IDLE)
+		    && !int_driven_rd_available(data)) {
+			/* Whole packet read, RX can be re-enabled. */
+			int_driven_rx_feed(dev, data);
+		}
+	}
+
+	return cpylen;
+}
+
+#if CONFIG_NRF_SW_LPUART_INT_DRIVEN
+
+static int api_fifo_fill(struct device *dev, const uint8_t *tx_data, int size)
+{
+	struct lpuart_data *data = get_dev_data(dev);
+	int err;
+
+	if (!atomic_cas((atomic_t *)&data->int_driven->txlen, 0, size)) {
+		return 0;
+	}
+
+	size = MIN(size, sizeof(data->int_driven->txbuf));
```

Comment:
`size` should be corrected before it is written to `data->int_driven->txlen`.

Reply:
done